Publish and Subscribe via Kafka

michaeldauria opened the pull request:
We use a ConsumerGroup on the subscriber side to allow for multiple consumers per topic.

michaeldauria committed:
Getting poseidon and poseidon_cluster in the door

michaeldauria committed:
Connect Poseidon and Poseidon Cluster

First pass, needs a refactor after the tests are passing

michaeldauria committed:
Publishing via Kafka works, no subscribing yet

michaeldauria committed:
basic_spec.rb:10 passing

michaeldauria committed:
Basic specs passing!

michaeldauria committed:
More tests passing

michaeldauria committed:
Use an appropriate topic key

michaeldauria committed:
Properly shut down DistributorThreads

michaeldauria committed:
Sleep less in DistributorThread fetch loops

michaeldauria committed:
Clearer logging for Distributor

michaeldauria committed:
sync STDERR

michaeldauria committed:
Better connections and disconnections to kafka

Commit:
Merge branch 'new-locking-scheme' into kafka

Conflicts:
	lib/promiscuous/config.rb
	lib/promiscuous/publisher/operation/base.rb

Commit:
Fix cold start spec

michaeldauria committed:
Add in Kafka macros

Commit:
Merge branch 'kafka' of github.com:promiscuous-io/promiscuous into kafka

Commit:
wip - consistent failures

Commit:
More specs passing

Commit:
Recovery specs passing

Commit:
ALL PASSING!!!

michaeldauria commented on lib/promiscuous/cli.rb:
```diff
@@ -71,7 +71,7 @@ def record
     endpoint = MultiJson.load(payload)['__amqp__']
     if endpoint
-      # TODO confirm
+      Promiscuous::Kafka.publish(:key => endpoint, :payload => payload)
```
this `:key` is incorrect

michaeldauria commented on lib/promiscuous/config.rb:
```diff
@@ -1,21 +1,26 @@
+                 :backend, :amqp_url, :publisher_amqp_url, :subscriber_amqp_url,
+                 :publisher_exchange, :subscriber_exchanges, :sync_exchange, :queue_name,
+                 :queue_options, :error_queue_name, :rabbit_mgmt_url, :queue_policy,
                  :error_exchange, :error_routing, :retry_routing, :error_ttl
```
`:error_ttl` will live on

michaeldauria commented on lib/promiscuous/config.rb:
```diff
@@ -1,21 +1,26 @@
   def self.backend=(value)
     @@backend = value
     Promiscuous::AMQP.backend = value
+    Promiscuous::Kafka.backend = (value == :bunny)? :poseidon : value
```
TODO: refactor to allow for either `bunny` or `kafka` for subscription and both for publishing

michaeldauria commented on lib/promiscuous/config.rb:
```diff
@@ -63,8 +77,8 @@ def self._configure(&block)
-    self.logger               ||= Logger.new(STDERR).tap { |l| l.level = ENV["LOGGER_LEVEL"].to_i || Logger::WARN }
-    self.subscriber_threads   ||= 10
+    self.logger               ||= defined?(Rails) ? Rails.logger : Logger.new(STDERR).tap { |l| l.level = Logger::WARN }
+    self.subscriber_threads   ||= 1
```
bump back up to 10?

kareemk replied:
Yes. And we should run tests with 10 to make sure everything still works.

michaeldauria commented on lib/promiscuous/kafka/poseidon.rb:
```diff
@@ -0,0 +1,149 @@
+
+class Promiscuous::Kafka::Poseidon
+
+  # TODO: Move to spec support
```
move to spec/support

michaeldauria commented on lib/promiscuous/kafka/poseidon.rb:
```diff
@@ -0,0 +1,149 @@
+                                          :max_send_retries => 10,
+                                          :retry_backoff_ms => 100,
+                                          :required_acks => 1,
+                                          :ack_timeout_ms => 1000,
```
most, if not all, of these need to be configurable

kareemk replied:
yes, and really understand the implications of each.

michaeldauria commented on lib/promiscuous/kafka/poseidon.rb:
```diff
@@ -0,0 +1,149 @@
+
+  def connect
+    @connection = new_connection
+    # Poseidon.logger = Logger.new(STDOUT).tap { |l| l.level = 0 }
```
move to spec/support

michaeldauria commented on lib/promiscuous/kafka/poseidon.rb:
```diff
@@ -0,0 +1,149 @@
+    end
+  end
+
+  #TODO: broken
```
broken

michaeldauria commented on lib/promiscuous/kafka/poseidon.rb:
```diff
@@ -0,0 +1,149 @@
+    10.times do
+      ok = @connection.send_messages([Poseidon::MessageToSend.new(options[:topic], options[:payload], options[:key])])
+      break if ok
+      sleep(1)
```
this might be too long to `sleep`

michaeldauria commented on lib/promiscuous/kafka/poseidon.rb:
```diff
@@ -0,0 +1,149 @@
+      end
+    end
+
+    def delete_queues
```
remove this method, we fast forward the topic instead

michaeldauria commented on lib/promiscuous/publisher/operation/base.rb:
```diff
@@ -101,15 +101,15 @@ def queue_operation_payloads(operations = self.operations)
       payload[:host]       = Socket.gethostname
+      payload[:key]        = lock_key
       payload.merge!(self.payload_attributes)
-      MultiJson.dump(payload)
```
if there is a way to get at this `lock_key` from the `instance` we can put this back and refactor

kareemk replied:
there's definitely a way to refactor this. i'll think about it.

michaeldauria commented on lib/promiscuous/subscriber/unit_of_work.rb:
```diff
@@ -40,8 +40,9 @@ def process_message
-      message.nack
+      # message.nack
       raise e if Promiscuous::Config.test_mode
+      sleep Promiscuous::Config.error_ttl / 1000.0; retry
```
use the `sleep` for Kafka and the `nack` for Rabbit

michaeldauria commented on lib/promiscuous/subscriber/worker/distributor.rb:
```diff
@@ -0,0 +1,89 @@
+      @stop = false
+      @topic = topic
+      @thread = Thread.new { main_loop }
+      # @thread.abort_on_exception = true
```
remove

michaeldauria commented on lib/promiscuous/subscriber/worker/distributor.rb:
```diff
@@ -0,0 +1,89 @@
+      @consumer = subscribe(@topic)
+      while not @stop do
+        begin
+          fetch_and_process_messages(&method(:on_message))
```
this should read `@consumer.fetch_and_process_messages` for clarity

michaeldauria commented on lib/promiscuous/subscriber/worker/distributor.rb:
```diff
@@ -0,0 +1,89 @@
+        begin
+          fetch_and_process_messages(&method(:on_message))
+        rescue Poseidon::Connection::ConnectionFailedError
+          @consumer = subscribe(@topic)
```
will this cause a rebalance and leave the replaced consumer in a broken state?

michaeldauria commented on lib/promiscuous/subscriber/worker/distributor.rb:
```diff
@@ -0,0 +1,89 @@
+      @thread.join
+    end
+
+    def show_stop_status(num_requests)
```
not sure where `num_requests` comes from

michaeldauria commented on lib/promiscuous/subscriber/worker/recorder.rb:
```diff
@@ -1,20 +1,13 @@
   def initialize(log_file)
     @log_file = log_file
-    extend Promiscuous::AMQP::Subscriber
+    extend Promiscuous::Kafka::Subscriber
```
we need to support both backends here

michaeldauria commented on spec/integration/consistency_spec.rb:
```diff
@@ -46,17 +46,17 @@
       eventually { SubscriberModel.count.should == 1 }
 
-      amqp_delayed!
+      kafka_delayed!
```
`backend_*`, since they function identically

michaeldauria commented on spec/integration/payload_spec.rb:
```diff
@@ -10,7 +10,7 @@
     PublisherModel.create(:field_1 => '1')
 
-    Promiscuous::AMQP::Fake.get_next_payload['host'].should == 'example.com'
+    Promiscuous::Kafka::Fake.get_next_payload['host'].should == 'example.com'
```
`Promiscuous::Backend::Fake`

michaeldauria commented on spec/integration/retry_exceptions_spec.rb:
```diff
@@ -38,7 +38,7 @@
 
     context 'when the ttl is longer then the subscriber is expected to receive the message' do
-      let(:ttl) { 99999 }
+      let(:ttl) { 2000 }
```
this number is actually in ms and not seconds

michaeldauria commented on spec/support/backend.rb:
```diff
@@ -27,6 +29,7 @@ def use_real_backend(options={}, &block)
         block.call(config) if block
       end
     end
+    Promiscuous::Kafka::Poseidon.advance_offsets_forward!
```
move this into support somewhere

michaeldauria commented on promiscuous.gemspec:
```diff
@@ -20,6 +20,8 @@ Gem::Specification.new do |s|
+  s.add_dependency "promiscuous-poseidon_cluster",    "~> 0.3.0 "
   s.add_dependency "ruby-progressbar",    "~> 1.2.0"
   s.add_dependency "redis",               "~> 3.0"
   s.add_dependency "algorithms",          "~> 0.6.1"
```
do we need this `algorithms` gem anymore?

benseligman replied:
did you try removing it?

michaeldauria committed:
Merge kafka

michaeldauria committed:
Minor config shifting around

Commit:
Deleting failures file

kareemk commented on lib/promiscuous.rb:
```diff
@@ -44,6 +44,7 @@ def configure(&block)
 
     def connect
       AMQP.connect
+      Kafka.connect # TODO: needed?
```
Figure this out.

kareemk commented on lib/promiscuous.rb:
```diff
@@ -51,6 +52,7 @@ def connect
 
     def disconnect
       AMQP.disconnect
+      Kafka.disconnect
```
Also is this needed?

michaeldauria replied:
this is yes, so we cleanly shutdown the connection to Kafka and ZK

kareemk commented on lib/promiscuous/kafka/file.rb:
```diff
@@ -0,0 +1,78 @@
+class Promiscuous::Kafka::File
```
file?

michaeldauria replied:
this is for the file backend which is what powers the recorder. Best bet here is to have a generic Backend class instead of one per driver which is what it is now

kareemk commented on lib/promiscuous/kafka/poseidon.rb:
```diff
@@ -0,0 +1,149 @@
+        partitions_path = "/consumers/#{topic}/offsets/#{topic}"
+        zk.children(partitions_path).each do |partition|
+          partition_offset_requests = [::Poseidon::Protocol::PartitionOffsetRequest.new(partition.to_i, -1, 1000)]
+          offset_topic_requests = [::Poseidon::Protocol::TopicOffsetRequest.new(topic, partition_offset_requests)]
```
Clean this all up somehow. It's very hard to follow.

kareemk commented on lib/promiscuous/kafka/poseidon.rb:
```diff
@@ -0,0 +1,149 @@
+  end
+
+  def new_connection
+    # TODO: client id needs to be unique across the system, use hostname+rand?
```
seems important

michaeldauria replied:
yea, i had taken care of this and forgot to remove the comment

kareemk commented on lib/promiscuous/kafka/poseidon.rb:
```diff
@@ -0,0 +1,149 @@
+
+  def raw_publish(options)
+    ok = false
+    10.times do
```
why?

michaeldauria replied:
we had taken this from that test suite to make sure that messages were being published. We should probably try to take it out and see how it behaves.

kareemk commented on lib/promiscuous/kafka/poseidon.rb:
```diff
@@ -0,0 +1,149 @@
+
+    def fetch_and_process_messages(&block)
+      Promiscuous.debug "[kafka] Fetching messages topic:#{@consumer.topic} #{Thread.current}"
+      @consumer.fetch(:commit => false) do |partition, payloads|
```
Overall you need to space out your lines more. Makes it much easier to read.

michaeldauria replied:
not sure what you mean, example? I thought we were supposed to line up all of the arguments with the first, which is what I had done here.

kareemk commented on lib/promiscuous/kafka/poseidon.rb:
```diff
@@ -0,0 +1,149 @@
+      end
+
+      # TODO: rename to advance_offset or commit?
+      def ack
```
probably ok as `ack` for now

michaeldauria replied:
It has to remain this way if we're going to have pluggable backends with the same interface

michaeldauria committed:
This should be in a spec helper

michaeldauria committed:
Some comment cleanup

michaeldauria committed:
Expose a way to adjust the Poseidon log level

michaeldauria committed:
Only try to send messages once

michaeldauria committed:
Properly pass in the topic to each thread

michaeldauria committed:
Better logging messages for the distributor

michaeldauria committed:
Lower the error_ttl default

michaeldauria committed:
Retry when we fail to process messages a limited number of times

Hopefully this can go away as it helps a particular
race condition when loading models within tests

michaeldauria committed:
Make the connected? method work, for now

michaeldauria committed:
Simplify subscribing, may need to add in better
sleep mechanisms depending on how it behaves in
production

michaeldauria committed:
Syncing is done on the main topics

michaeldauria committed:
Pluggable backend! (one at a time for now)

Next refactor will include:
- rename Bunny -> AMQP and Poseidon to Kafka
- dependency injection for process_message errors
- only start/connect what we need per driver
  for subscriber/worker.rb

michaeldauria commented on .travis.yml:
```diff
@@ -1,11 +1,13 @@
-env: TIMEOUT=10
+env:
+  - TIMEOUT=10
+  - PROMISCUOUS_BACKEND=bunny
```
this will become poseidon once i can get my own Kafka cluster started and the tests running on it

michaeldauria commented on lib/promiscuous/subscriber/unit_of_work.rb:
```diff
@@ -36,12 +36,25 @@ def self.current=(value)
-      message.nack
       raise e if Promiscuous::Config.test_mode
+
+      # TODO: use dependency injection here instead
```
or some other cleaner approach

benseligman replied:
i like the approach we were talking about of moving `process_message` into each backend

kareemk replied:
this has to move out. yes, a process_message in backend makes sense.

michaeldauria committed:
Fix the recorder

benseligman commented on lib/promiscuous/backend/poseidon.rb:
```diff
@@ -0,0 +1,128 @@
+                                          :required_acks => 1,
+                                          :ack_timeout_ms => 1000,
+                                          :socket_timeout_ms => Promiscuous::Config.socket_timeout)
+    @connection
```
not necessary -- the assignment will return the value

benseligman commented on lib/promiscuous/backend/poseidon.rb:
```diff
@@ -0,0 +1,128 @@
+
+  # TODO: extend Poseidon with a connected? method
+  def connected?
+    !@connection.nil?
```
`@connection.present?`

benseligman commented on lib/promiscuous/backend.rb:
```diff
@@ -0,0 +1,50 @@
+
+    def driver=(value)
+      disconnect
+      @driver_class = value.nil? ? nil : "Promiscuous::Backend::#{value.to_s.camelize.gsub(/backend/, 'Backend')}".constantize
```
not a big deal, but i find `value.try { |v| "Promiscuous::Backend::#{value.to_s.camelize.gsub(/backend/, 'Backend')}".constantize }` a little easier to read

michaeldauria replied:
agreed

kareemk replied:
you shouldn't have to specify namespace as its in the same namespace. also that gsub is dirty. remove backend completely.

benseligman commented on lib/promiscuous/backend/poseidon.rb:
```diff
@@ -0,0 +1,128 @@
+      end
+    end
+
+    module Worker
```
just noticing that the worker just delegates to the distributor -- is there a need for both?

benseligman commented on lib/promiscuous/subscriber/worker/distributor.rb:
```diff
@@ -0,0 +1,88 @@
+  def start
+    num_threads = Promiscuous::Config.subscriber_threads
+    Promiscuous::Config.subscriber_topics.each do |topic|
+      @distributor_threads ||= num_threads.times.map { DistributorThread.new(topic) }
```
why is this cached?

michaeldauria replied:
In case you call start twice by mistake, i guess that's defensive coding but i took the idea from the runner

benseligman commented on lib/promiscuous/subscriber/worker/distributor.rb:
```diff
@@ -0,0 +1,88 @@
+
+  class DistributorThread
+    def initialize(topic)
+      # late include of CelluloidSubscriber because the class is resolved
```
haha thanks for the comment -- it took me and @TomHoss a while to understand why another extend in promiscuous was inside initialize

benseligman commented on spec/integration/amqp_spec.rb:
```diff
@@ -1,11 +1,13 @@
 
-describe Promiscuous::AMQP do
-  before { use_real_backend }
+if Promiscuous::Config.backend == :bunny
```
why this conditional in tests?

benseligman replied:
nvm i understand

michaeldauria committed:
Fix Travis ENV variables

kareemk commented on .travis.yml:
```diff
@@ -1,11 +1,13 @@
-env: TIMEOUT=10
+env:
+  - TIMEOUT=10 PROMISCUOUS_BACKEND=bunny
+  - TIMEOUT=10 PROMISCUOUS_BACKEND=poseidon
```
BACKEND instead of PROMISCUOUS_BACKEND

michaeldauria committed:
Ben comments

michaeldauria committed:
Fix the recorder

michaeldauria committed:
Wire up the replay correctly

kareemk commented on lib/promiscuous/backend/poseidon.rb:
```diff
@@ -0,0 +1,128 @@
+
+  def raw_publish(options)
+    ok = @connection.send_messages([Poseidon::MessageToSend.new(options[:topic], options[:payload], options[:key])])
+    raise "Unable to send messages" if !ok
```
needs to be a Promiscuous::Error subclass

kareemk commented on lib/promiscuous/backend/poseidon.rb:
```diff
@@ -0,0 +1,128 @@
+  def publish(options={})
+    @connection_lock.synchronize do
+      raw_publish(options)
+      # TODO: Rabbit is in control of the unlocking for now
```
how is this working then in tests?

michaeldauria replied:
This was a bad comment and will be addressed with the `:both` backend i am working on

kareemk commented on lib/promiscuous/subscriber/worker.rb:
```diff
@@ -1,33 +1,37 @@
-    @pump = Pump.new(self)
-    @runner = Runner.new(self)
+    # inject what we need for our backend
+    extend Promiscuous::Backend.subscriber_worker_module
```
subscriber_methods

kareemk commented on spec/support/backend.rb:
```diff
@@ -1,5 +1,41 @@
+require 'securerandom'
+
 module BackendHelper
+  def advance_offsets_forward!
```
this should be in KafkaHelper

kareemk commented on spec/support/backend.rb:
```diff
@@ -13,24 +49,37 @@ def reconfigure_backend(&block)
       config.destroy_check_interval = 0
       config.max_retries = 0
       config.rabbit_mgmt_url = rabbit_mgmt_url
+      # config.kafka_hosts = kafka_hosts
```
why commented out?

kareemk commented on spec/support/backend.rb:
```diff
@@ -73,14 +122,34 @@ def amqp_url
     ENV['BOXEN_RABBITMQ_MGMT_URL'] || 'http://guest:guest@localhost:15672'
   end
+
+  def kafka_hosts
```
you should split out the helpers for rabbit vs. kafka.

michaeldauria committed:
PROMISCUOUS_BACKEND -> BACKEND

michaeldauria committed:
subscriber_worker_module -> subscriber_methods

michaeldauria committed:
Proper error handling for failed messages

michaeldauria committed:
Create separate backend helpers

michaeldauria committed:
Pull out process_message into each backend

benseligman commented:
lgtm

michaeldauria committed:
Error handling is already done within publish

michaeldauria committed:
Add in a Both backend

Will publish to both and subscribe only to Rabbit

michaeldauria committed:
Publish messages on the correct key

michaeldauria committed:
rabbit -> backend

michaeldauria committed:
Add in some logging for when we unlock

michaeldauria committed:
Properly clean up while running the both backend

michaeldauria committed:
Catch the UnableToFetchMetadata exception

michaeldauria committed:
A message may come in already parsed

michaeldauria committed:
Return the parsed payload as a hash with
indifferent access if already a hash.

This lines up correctly with the way MultiJson.load
works and ensures that properties are accessible 
using the string syntax.

michaeldauria committed:
Improve debug logging for Kafka

michaeldauria committed:
When we’re unable to fetch metadata, retry a
handful of times before bubbling up the error

michaeldauria committed:
Increase the socket_timeout default from 10 -> 50ms

michaeldauria committed:
Boot and use our own Kafka & Zookeeper

michaeldauria committed:
Rescue from StandardError, not Exception

michaeldauria committed:
Let’s keep the interface the same and use `ack`

michaeldauria committed:
Report all backends when there is an exception connecting

michaeldauria committed:
When pulling in code for injection, use the lookup
chain instead of strings

michaeldauria committed:
Run Travis on the latest Ruby versions

michaeldauria committed:
Wait for Kafka to be ready rather than blindly sleeping

michaeldauria committed:
Change to strings so that ActiveRecord tests are doing the correct comparison

michaeldauria committed:
Only start Kafka when needed

michaeldauria committed:
Fix a bug where we were unable to subscribe to more than one topic

michaeldauria committed:
Properly track threads in the Distributor

michaeldauria committed:
Ruby 2.1.5

michaeldauria committed:
Use a more appropriate lock scope for subscribers

michaeldauria committed:
Default to just 1 subscriber thread

michaeldauria committed:
Break out the options we pass to Poseidon::ConsumerGroup

michaeldauria committed:
Use blank? instead of empty?

michaeldauria committed:
Use our release of poseidon_cluster with bugfixes

michaeldauria committed:
Advance the correct topic

michaeldauria committed:
Ensure we only save one model

michaeldauria committed:
Move error_retry_max to Config and lower to 10

michaeldauria committed:
Retry spec passing reliably

michaeldauria committed:
Use the correct namespace when recovering locks

michaeldauria committed:
Use more appropriate socket timeout options

michaeldauria committed:
Need to bring back Subscriber::UnitOfWork#app for instrumentation

michaeldauria committed:
Default backend is :poseidon

michaeldauria committed:
Commit our offset once per fetch rather than once per payload

michaeldauria committed:
Sync over a separate topic

michaeldauria committed:
Update poseidon_cluster for bug fixes

michaeldauria committed:
Sync to the target instead of the local app

michaeldauria committed:
Lower the retry thresholds

michaeldauria committed:
Increase the claim timeout

Gives more time for consumers to finish up their
work before giving up their previous claim.

michaeldauria committed:
Lower the size of a fetch to get less messages

This helps when consumers are far behind and new
consumer join the cluster

michaeldauria committed:
It’s OK that a lock was recovered

michaeldauria committed:
Bug fix: Do not publish when a Mongo operation fails

michaeldauria committed:
Only index the publishers

michaeldauria committed:
Update poseidon_cluster

michaeldauria committed:
Upgrade poseidon_cluster

michaeldauria committed:
Pass payload into `error_notifier`

michaeldauria committed:
Bugfix: Connect once we have set a backend

benseligman committed:
Better error when publishing a dirty model

Previously, the error message made it seem like you could refresh the instance
in memory. This should make it clearer that the instance needs to be refetched
from the database.

signed-off-by michaeldauria

michaeldauria committed:
Update max_wait_ms to 50ms

michaeldauria committed:
Up the max_bytes to 600Kb

michaeldauria committed:
Upgrade to robust-redis-lock ~> 1.3.0

benseligman committed:
Ephemeral saves do not raise when non-instrumented

Previously, inside a without_promiscuous block, ephemeral persistence failed.
Since the whole point of ephemerals and there is no other event necessary, this
change makes ephemeral noninstrumented saves a noop.

signed-off-by michaeldauria
